apps/gdalalg_main.cpp

```diff
@@ -0,0 +1,123 @@
+/******************************************************************************
+*
+* Project:  GDAL
+* Purpose:  gdal "main" command
+* Author:   Even Rouault <even dot rouault at spatialys.com>
+*
+******************************************************************************
+* Copyright (c) 2024, Even Rouault <even dot rouault at spatialys.com>
+*
+* SPDX-License-Identifier: MIT
+****************************************************************************/
+
+#include "gdalalg_main.h"
+
+//! @cond Doxygen_Suppress
+
+/************************************************************************/
+/*                  GDALMainAlgorithm::GDALMainAlgorithm()              */
+/************************************************************************/
+
+GDALMainAlgorithm::GDALMainAlgorithm()
+    : GDALAlgorithm(NAME, DESCRIPTION, HELP_URL)
+{
+    for (const std::string &subAlgName :
+         GDALGlobalAlgorithmRegistry::GetSingleton().GetNames())
+    {
+        const auto pInfo =
+            GDALGlobalAlgorithmRegistry::GetSingleton().GetInfo(subAlgName);
+        if (pInfo)
+            RegisterSubAlgorithm(*pInfo);
+    }
+
+    m_longDescription = "'gdal <FILENAME>' can also be used as a shortcut for "
+                        "'gdal info <FILENAME>'.\n"
+                        "And 'gdal read <FILENAME> ! ...' as a shortcut for "
+                        "'gdal pipeline <FILENAME> ! ...'.";
+
+    SetDisplayInJSONUsage(false);
+}
+
+/************************************************************************/
+/*              GDALMainAlgorithm::ParseCommandLineArguments()          */
+/************************************************************************/
+
+bool GDALMainAlgorithm::ParseCommandLineArguments(
+    const std::vector<std::string> &args)
+{
+    if (args.size() >= 2 && args[0] == "read")
+    {
+        m_subAlg =
+            GDALGlobalAlgorithmRegistry::GetSingleton().Instantiate("pipeline");
+        if (m_subAlg)
+        {
+            bool ret = m_subAlg->ParseCommandLineArguments(args);
+            if (ret)
+            {
+                m_selectedSubAlg = &(m_subAlg->GetActualAlgorithm());
+                std::vector<std::string> callPath(m_callPath);
+                callPath.push_back("vector");
+                m_selectedSubAlg->SetCallPath(callPath);
+                return true;
+            }
+            else if (strstr(CPLGetLastErrorMsg(),
+                            "has both raster and vector content"))
+            {
+                m_showUsage = false;
+                return false;
+            }
+        }
+    }
+    else if (!(args.size() >= 1 && InstantiateSubAlgorithm(args[0])))
```

Comment:
Comments added and code reworked as:
```
    // Detect shortest form of pipeline:
    // "gdal read in.tif ! .... ! write out.tif"
    if (args.size() >= 2 && args[0] == "read")
    {
      ...
    }
    // Generic case: "gdal {subcommand} arguments"
    // where subcommand is a known subcommand
    else if (args.size() >= 1 && InstantiateSubAlgorithm(args[0]))
    {
        return GDALAlgorithm::ParseCommandLineArguments(args);
    }
    // Otherwise check if that is the shortest form of "gdal read mydataset"
    // where "read" is omitted: "gdal in.tif"
    else
    {
        ....
    }
```